.grit/patterns/python/warning-for-tempfile-without-flush.md

```diff
@@ -0,0 +1,227 @@
+---
+title: Warning for tempfile without flush
+---
+
+"Be cautious when using `$F.name` without preceding it with `.flush()` or `.close()`, as it may result in an error. This is because the file referenced by `$F.name` might not exist at the time of use. To prevent issues, ensure that you either call `.flush()` to write any buffered data to the file or close the file with .close() before referencing `$F.name`.
+
+- [reference](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp)
+
+tags: #fix #good-practice
+
+```grit
+engine marzano(0.1)
+language python
+
+or {
+    `with $tempfile.NamedTemporaryFile($params) as $f:` as $fileOpen where {
+        and {
+            $fileOpen <: not contains `$f.flush()`,
+            $fileOpen <: not contains `$f.close()`,
+            $fileOpen <: contains `$f.name`,
+            $fileOpen => `# BAD: should add $f.flush() or $f.close() before calling $f.name \n $fileOpen`
```

Comment:
Don't manually add annotations like this via rewrites. Grit already has support for checking things - if no meaningful fix is available, you can skip having a rewrite at all.

In this case, a *meaningful* autofix might be inserting f.close() on the line before the f.name() call.

Reply:
Hi @morgante 

> In this case, a meaningful autofix might be inserting f.close() on the line before the f.name() call.

Yes, true I am trying to build the same.

any suggestion, if to build this, 

`cmd = [binary_name, fout.name, *[str(path) for path in targets]]`

parsing this and add `$f.close()` before this, become difficult

Reply:
You don't need to handle all cases. It's fine to skip any that are too complex.

For inserting things it's often best to find the containing statement() and append to that.